Sources/JavaScriptKit/JSFunction.swift

```diff
@@ -1,45 +1,44 @@
-    public func dynamicallyCall(withArguments arguments: [JSValueConvertible]) -> JSValue {
-        let result = arguments.withRawJSValues { rawValues -> RawJSValue in
+    func callAsFunction(this: JSObjectRef? = nil, args: [JSValueConvertible]) -> JSValue {
+        let result = args.withRawJSValues { rawValues in
             rawValues.withUnsafeBufferPointer { bufferPointer -> RawJSValue in
                 let argv = bufferPointer.baseAddress
                 let argc = bufferPointer.count
                 var result = RawJSValue()
-                _call_function(
-                    self.id, argv, Int32(argc),
-                    &result.kind, &result.payload1, &result.payload2, &result.payload3
-                )
+                if let thisId = this?.id {
+                    _call_function_with_this(thisId,
+                                             self.id, argv, Int32(argc),
+                                             &result.kind, &result.payload1, &result.payload2, &result.payload3)
+                } else {
+                    _call_function(
+                        self.id, argv, Int32(argc),
+                        &result.kind, &result.payload1, &result.payload2, &result.payload3
+                    )
+                }
                 return result
             }
         }
         return result.jsValue()
     }
 
-    public func apply(this: JSObjectRef, arguments: JSValueConvertible...) -> JSValue {
-        apply(this: this, argumentList: arguments)
+    @discardableResult
+    public func callAsFunction(this: JSObjectRef? = nil, _ args: JSValueConvertible...) -> JSValue {
+        self(this: this, args: args)
     }
 
-    public func apply(this: JSObjectRef, argumentList: [JSValueConvertible]) -> JSValue {
-        let result = argumentList.withRawJSValues { rawValues in
-            rawValues.withUnsafeBufferPointer { bufferPointer -> RawJSValue in
-                let argv = bufferPointer.baseAddress
-                let argc = bufferPointer.count
-                var result = RawJSValue()
-                _call_function_with_this(this.id,
-                                         self.id, argv, Int32(argc),
-                                         &result.kind, &result.payload1, &result.payload2, &result.payload3)
-                return result
-            }
-        }
-        return result.jsValue()
+    public func new(_ args: JSValueConvertible...) -> JSObjectRef {
+        new(args: args)
     }
 
-    public func new(_ arguments: JSValueConvertible...) -> JSObjectRef {
-        return arguments.withRawJSValues { rawValues in
+    // Guaranteed to return an object because either:
+    // a) the constructor explicitly returns an object, or
+    // b) the constructor returns nothing, which causes JS to return the `this` value, or
+    // c) the constructor returns undefined, null or a non-object, in which case JS also returns `this`.
+    public func new(args: [JSValueConvertible]) -> JSObjectRef {
```

Comment:
```suggestion
    public func new(arguments: [JSValueConvertible]) -> JSObjectRef {
```

I think there shouldn't be abbreviations in public interface. 🙏